2.train-VAE/vae.py

```diff
@@ -0,0 +1,150 @@
+            - K.square(self.encoder_block["z_mean"])
+            - K.exp(self.encoder_block["z_log_var"])
+        )
+        self.kl_loss = K.sum(self.kl_loss, axis=-1)
+        self.kl_loss *= -0.5  
+        
+        batch_size = K.shape(self.encoder_block["z"])[0]
+        latent_dim = K.int_shape(self.encoder_block["z"])[1]
+        true_samples = K.random_normal(shape=(batch_size, latent_dim), mean=0., stddev=1.)
+        self.mmd_loss = compute_mmd(true_samples, self.encoder_block["z"])
+        
+        mmd_loss = K.get_value(self.lam)* self.mmd_loss
+        kl_loss = K.get_value(self.beta)* self.kl_loss
+#         mmd_loss = kl_loss
+        total_loss = self.reconstruction_loss + mmd_loss + kl_loss
+#         total_loss = self.reconstruction_loss + kl_loss
+
+ 
+        return {
+            "loss": total_loss,
+            "reconstruction_loss": self.reconstruction_loss,
+            "kl_loss": kl_loss,
+            "mmd_loss": mmd_loss,
+        }
+    
+    def compile_encoder(self, name="encoder"):
+        self.encoder_block = connect_encoder(
+            input_dim=self.input_dim,
+            latent_dim=self.latent_dim,
+            architecture=self.encoder_architecture,
+            batch_norm=self.encoder_batch_norm,
+        )
+        self.inputs = self.encoder_block["inputs"]
+
+    def compile_decoder(self, name="decoder"):
+        self.decoder_block = connect_decoder(
+            input_dim=self.input_dim,
+            latent_dim=self.latent_dim,
+            architecture=self.decoder_architecture,
+        )
+
+    def compile_vae(self):
+       
+        self.compile_encoder()
+        self.compile_decoder()
+        self.setup_optimizer()
+
+        # instantiate VAE model
+        self.cycle = self.decoder_block["decoder"](
+            self.encoder_block["encoder"](self.inputs)[2]
+        )
+        self.vae = Model(self.inputs, self.cycle, name="vae_mlp")
+
+        self.vae_loss = self.compile_loss()['loss']
+        kl_loss = self.compile_loss()['kl_loss']
+        mmd_loss = self.compile_loss()['mmd_loss']
+        reconstruction_loss = self.compile_loss()['reconstruction_loss']
+        self.vae.add_loss(self.vae_loss)
+        self.vae.add_metric(reconstruction_loss, aggregation='mean', name='recon')
+        self.vae.add_metric(kl_loss, aggregation='mean', name='kl')
+        self.vae.add_metric(mmd_loss, aggregation='mean', name='mmd')
+        self.vae.compile(optimizer=self.optim)
+
+    def setup_optimizer(self):
+        if self.optimizer == "adam":
+            self.optim = optimizers.Adam(lr=self.learning_rate)
+
+    def train(self, x_train, x_test):
+        if not hasattr(self, "vae_loss"):
+            self.compile_vae()
+        # train the autoencoder
+        self.vae.fit(
+            x_train,
+            epochs=self.epochs,
+            batch_size=self.batch_size,
+            validation_data=(x_test, None),
+            verbose=self.verbose,
+        )
```

Comment:
```suggestion
        )

```

.py files need an extra new line at the EOF